core/src/toga/widgets/selection.py

```diff
@@ -1,100 +1,155 @@
+                    "`on_select` has been deprecated, use `on_change`"
+                )
+            else:
+                warnings.warn(
+                    "Selection.on_select has been renamed Selection.on_change"
+                )
+                on_change = on_select
+
+        self.on_change = on_change
         self.enabled = enabled
 
     @property
-    def items(self):
-        """The list of items.
+    def items(self) -> ListSource:
+        """The list of items to display in the selection, as a ListSource.
+
+        A value of None is turned into an empty ListSource.
+
+        A literal list of values will be converted into a ListSource. The
+        conversion that takes place depends on the type of data in the list:
 
-        Returns:
-             The ``list`` of ``str`` of all selectable items.
+         * Literal strings will be used as-is.
+
+         * Non-iterable objects will be converted to a string for display
+           purposes, but will be returned in their original form.
+
+         * If the object is iterable, the value of the attribute described
+           by the ``accessor`` argument provided at the time of construction
+           will be used. If no ``accessor`` was provided, the ListSource will
+           look for an attribute named ``value``.
         """
         return self._items
 
     @items.setter
     def items(self, items):
-        self._impl.remove_all_items()
-        self._items = items
+        if self._accessor is None:
+            accessors = ["value"]
+        else:
+            accessors = [self._accessor]
 
-        for i in items:
-            self._impl.add_item(i)
+        if items is None:
+            self._items = ListSource(accessors=accessors, data=[])
+        elif isinstance(items, (list, tuple)):
+            self._items = ListSource(accessors=accessors, data=items)
+        else:
+            if self._accessor is None:
+                raise ValueError("Must specify an accessor to use a data source")
+            self._items = items
+
+        self._items.add_listener(self._impl)
+        self._impl.change_source(source=self._items)
+        self.refresh()
+
+    def _title_for_item(self, item):
+        """Internal utility method; return the display title for an item"""
+        if self._accessor:
+            title = getattr(item, self._accessor)
+        else:
+            title = item.value
+
+        return str(title)
 
     @property
     def value(self):
-        """The value of the currently selected item.
+        """The currently selected item.
 
-        Returns:
-            The selected item as a ``str``.
+        Returns None if there are no items in the selection.
+
+        If changing the current value, ValueError is raised if the specified
+        item cannot be found in the data source.
         """
-        return self._impl.get_selected_item()
+        item = self._impl.get_selected_item()
```

Comment:
To reduce duplication between the implementations, it would be simpler if this was just a `get_selected_index` method.

Reply:
Good catch on `get_selected_index()`; I didn't notice the duplication, so I've made that change.

As for making the implementation the listener - I definitely see the appeal with that. One complication is iOS, as you've noted; the other is collision with the API on the interface. `insert`/`remove`/`clear` already exist as methods there, controlling the addition of children, so at the very least, we'd need to introduce a proxy class specifically to act as the listener; and for most implementations, all we'd really be abstracting is (a) access to `interface.items`, and (b) converting an item to a string.